isy-security/src/main/java/de/bund/bva/isyfact/security/oauth2/client/IsyOAuth2Authentifizierungsmanager.java

```diff
@@ -2,89 +2,152 @@
 
 import org.springframework.lang.Nullable;
 import org.springframework.security.authentication.AuthenticationProvider;
+import org.springframework.security.authentication.BadCredentialsException;
 import org.springframework.security.authentication.ProviderManager;
 import org.springframework.security.core.Authentication;
 import org.springframework.security.core.AuthenticationException;
 import org.springframework.security.core.context.SecurityContextHolder;
 import org.springframework.security.oauth2.client.registration.ClientRegistration;
 import org.springframework.security.oauth2.client.registration.ClientRegistrationRepository;
+import org.springframework.security.oauth2.client.registration.ClientRegistrations;
 import org.springframework.security.oauth2.core.AuthorizationGrantType;
 import org.springframework.util.Assert;
 
-import de.bund.bva.isyfact.security.oauth2.client.authentication.ClientCredentialsAuthenticationToken;
-import de.bund.bva.isyfact.security.oauth2.client.authentication.ManualClientCredentialsAuthenticationToken;
-import de.bund.bva.isyfact.security.oauth2.client.authentication.PasswordAuthenticationToken;
+import de.bund.bva.isyfact.security.config.IsyOAuth2ClientConfigurationProperties;
+import de.bund.bva.isyfact.security.config.IsyOAuth2ClientConfigurationProperties.AdditionalRegistrationProperties;
+import de.bund.bva.isyfact.security.oauth2.client.authentication.token.ClientCredentialsClientRegistrationAuthenticationToken;
+import de.bund.bva.isyfact.security.oauth2.client.authentication.token.ClientCredentialsRegistrationIdAuthenticationToken;
+import de.bund.bva.isyfact.security.oauth2.client.authentication.token.PasswordClientRegistrationAuthenticationToken;
 
 /**
  * Default implementation of the {@link Authentifizierungsmanager} that should suffice for most use cases.
  * <p>
  * It provides different ways to authorize OAuth 2.0 Clients (via the Client Credentials flow) and
  * System users (representing resource owners) via the Resource Owner Password Credentials flow.
+ * <p>
+ * The primary way for authentication is {@link #authentifiziere(String)}, which depends on OAuth 2.0 Client Registrations
+ * to be configured in the application properties.
+ * The other {@code authentifiziereClient}/{@code authentifiziereSystem} methods construct the necessary
```

Comment:
the three methods `authentifiziere`, `authentifiziereClient` and `authentifiziereSystem` have no priority nor a preference in their usage.
There is a difference in the invocation-context:
- **internal clientRegistration:** the authentification data are defined internally as _clientRegistrations_ in local configuration files. This is only allowed for _confidential clients_, which are categorized secure enough to take care about their secrets in their local environment and in their own responsebility. _Confidential clients_ will authenticate themselfs by using the method `authentifiziere`, which will internally decide which credential flow it will use. The decision base on the configured _clientRegistration_ property `grant-type`.
- **external clientRegistration:** the authentication data is send from an external Client with each request and the infrastructure has to authenticate these requests. The first infrastructure-component of such an external request will be a _service-gateway_ or in the future an _authentication-gateway_. These gateways act as an _intermediary_ and have to authenticate themselfs during the authentication process against the IAM-service. So both methods `authentifiziereClient` and `authentifiziereSystem` are use by these gateways. The method `authentifiereSystem` has the special requirement within the _resource-owner-credential-flow_ to authenticate the intermediary too, which the _client-credential-flow_ not have.

So there is no priority about the usage of one of these three authentication methods. The only preference is to use the _client-credential flow_ than the _resource-owner-credential-flow_, but this depends on the migration state of the users authentication data in IAM, if he is still registered as a _natural user_ or as a _technical client_.

Reply:
see above